enhancements/machine-api/control-plane.md

```diff
@@ -0,0 +1,335 @@
+
+#### Declarative horizontal scaling
+##### Scale out
+1. The controller always reconciles towards expected number of replicas. This must be an odd number.
+2. Fetch all existing control plane Machine resources by ownerRef. Adopt any other machine having a targeted label e.g `node-role.kubernetes.io/master`.
+3. Compare with expected replicas number. If expected is higher than current then:
+4. Check all owned machines have a backed ready node.
+5. Check all etcd members for all owned machines are healthy via Cluster etcd Operator status signalling.
+6. If (NOT all etcd members are healthy OR NOT all owned machines have a backed ready node) then controller short circuits here, log, update status and requeue. Else:
+7. Choose a failure domain.
+8. Create new Machine object with a templated spec. Go to 1.
+9. Cluster etcd Operator watches the new node. It runs a new etcd pod on it.
+
+##### Scale in
+1. The controller always reconciles towards expected number of replicas. This must be an odd number.
+2. Fetch all existing control plane Machine resources by ownerRef. Adopt any other machine having a targeted label e.g `node-role.kubernetes.io/master`.
+3. Compare with expected replicas number. If expected is lower than current then:
+4. Check all owned machines have a backed ready node.
+5. Check all etcd members for all owned machines are healthy via Cluster etcd Operator status signalling.
+6. If (NOT all etcd members are healthy OR NOT all owned machines have a backed ready node) then controller short circuits here, log, update status and requeue. Else:
+7. Pick oldest machine in more populated failure domain.
+8. Remove etcd member.
+9. Delete machine. Go to 1. (Race between the node going away and Cluster etcd Operator re-adding the member?)
+
+#### Declarative Vertical scaling (scale out + scale in)
+1. Watch changes to the Control Plane provideSpec
+2. Fetch all existing control plane Machine resources by ownerRef. Adopt any other machine having a targeted label.
+3. Fetch all machines with old providerSpec.
+4. If any machine has old providerSpec then signal controller as "needs upgrade".
+5. If any machine has "replaced" annotation then trigger scale in workflow (starting in 4) and requeue. Else:
+6. Pick oldest machine in more populated failure domain.
+7. Trigger scale out workflow (starting in 4). Set "replaced" annotation. Requeue.
+This is effectively a rolling upgrade with maxUnavailable 0 and maxSurge 1.
+
+#### Self healing (MHC + reconciling)
+1. The Controller should always reconcile by removing etcd members for voluntary Machine disruptions, i.e machine deletion.
+2. At creation time, unless indicated otherwise `EnableAutorepair=true` will be default. The controller will create a Machine Health Checking resource with an ownerRef which will monitor the Control Plane Machines. This will request unhealthy machines to be deleted. Related https://github.com/openshift/machine-api-operator/pull/543.
```

Comment:
Are we going to support rebooting instead of or in addition to reprovisioning for bare metal clusters?

Reply:
This proposal supports and is considering remediating through deletion only. At some point I'd expect a proposal PR against MHC to flesh out the reboot/baremetal strategy if that's something we want to fully support in the product.

Reply:
There is already [work being done to support alternative remediation methods](https://github.com/openshift/cluster-api-provider-baremetal/pull/59 ). I think we need to incorporate that into this design to maintain consistent support across platforms, or be explicit in saying that we're not going to support self-healing the control plane for bare metal.

Maybe @beekhof or @n1r1 can provide more background.

Reply:
I agree, I'm aware of work being done. We enabled exploring external remediation through an annotation. Now for that to be supported we need a RFE PR following up the original MHC fleshing out e2e details. This or any other design consuming MHC shouldn't be tied to that until there's a fleshed story in place. Clarified https://github.com/openshift/enhancements/pull/278/commits/0fa380bc1e8d1f625ae7a7a11ac56adb153c6c50

Reply:
The clarification is an improvement, but I would still like to see an explicit statement that it is not supported on bare metal. We're all striving to make metal work like cloud, and I'm concerned that there's an assumption that it does work the same unless otherwise stated. We keep surprising PMs and field folks when they find cases where we haven't clearly documented lack of support.